### PR TITLE
GH-2497 fix RDF collection processing for malformed collection structures

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #2497  

Briefly describe the changes proposed in this PR:

- Fix handling of malformed collection structures by checking beforehand, and falling back to serializing as regular RDF statements if the collection is not well-formed
- Added several regression test cases for various kinds of malformedness.

---- 
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by hmottestad** [comment]
@jeenbroekstra I found another blank node inlining bug. Maybe you want to take a look since you fixed the last one.

**Comment by jeenbroekstra** [comment]
Bit hard to tell where the discrepancy is from this. Can you narrow down where in the example input things get mapped wrong?

**Comment by hmottestad** [comment]
Narrowed it down and also used Jena to cleanup the input data somewhat.

File: core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java [hunk 1]
```diff
@@ -194,4 +194,170 @@ public void testNoBuffering() throws Exception {
...
+		String data = "" +
+				"@prefix ex:    <http://example.com/ns#> .\n" +
+				"@prefix sh:    <http://www.w3.org/ns/shacl#> .\n" +
+				"@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n" +
+				"@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .\n" +
+				"@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .\n" +
+				"@prefix foaf:  <http://xmlns.com/foaf/0.1/> .\n" +
+				"\n" +
+				"ex:a  ex:list   (_:b0 _:b0) .";
+
```
Comment (hmottestad) [inline core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:344]: This seems to be a representative example.

Comment (hmottestad) [reply core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:344]: This is what RDF4J outputs:

```
@prefix ex: <http://example.com/ns#> .
@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .

ex:a ex:list ([] []) .
```

Comment (jeenbroekstra) [reply core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:344]: ...that is correct, as far as I can tell. In this case, the bug might actually be in the isomorphism detection, because as far as I can tell those two graphs _should_ be isomorphic.

Comment (jeenbroekstra) [reply core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:344]: Then again, what we're actually after here is no loss of information of course. I'll take a closer look at the writer, see why this node gets inlined when it's referred to more than once.

**Comment by jeenbroekstra** [comment]
I've pushed a partial fix that addresses the simplified use case - unfortunately the more complex one still fails. The crux is probably more generally the list handling (which is a separate code path from how other statements with blank nodes are inlined, hence this being a corner case).

**Comment by hmottestad** [comment]
That's great @jeenbroekstra . I'm glad I kept the more complex one, but it's really hard to debug so only really useful as a validation and not for figuring out what's going on.

**Comment by jeenbroekstra** [comment]
I think we can isolate further by creating a few more small list-specific corner cases. We'll keep the original test until we're happy the corner cases cover all the problems.

File: core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java [hunk 2]
```diff
@@ -194,4 +194,196 @@ public void testNoBuffering() throws Exception {
...
+	@Test
+	public void anotherBnodeTestReduced2() throws Exception {
+		String data = "@prefix ex:    <http://example.com/ns#> .\n" +
+				"@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .\n" +
+				"\n" +
+				"ex:ex1    ex:list  [ rdf:first  [  ] ;\n" +
+				"                             rdf:rest   rdf:nil ;\n" +
+				"                             ex:abc   ex:def\n" +
+				"                           ]\n" +
```
Comment (jeenbroekstra) [inline core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:367]: I am not sure this one really counts as a bug, to be honest - and even if it does it is a _very_ exotic corner case. 

This case is an invalid collection structure in the sense that it links a non-list related poperty (`ex:abc`) to a bnode that that is directly part of the list structure itself. We are not so much inlining that blank node as just removing it, because the list gets converted to a syntactical shortform where the blank nodes and (rdf:first/rest properties) are just erased.

The actual output for this is:

```
@prefix ex: <http://example.com/ns#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .

ex:ex1 ex:list ([]) .
[] ex:abc ex:def .
```

In other words, one list with a single item (a blank node), and a statement with a blank node subject. The fact that in the original file that blank node was the _same_ blank node as the one used in the internal collection structure is lost, but that does not make the output incorrect. There's also a strong whiff of "what the heck are you even _doing_ using a collection blank node like this?!"

We could potentially handle this with a separate test to check if the collection is well-formed before serializing it, but it's going to be costly.

Comment (hmottestad) [reply core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:367]: I know it's pretty funky. I have no idea how it got produced to be honest. I still think we should consider it a bug though.

Comment (hmottestad) [reply core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:367]: I think the output is incorrect. The pretty printer and bnode inliner should return the same data if you first output n-triples, then pretty print and inline it, then convert it to n-triples again.

**Comment by hmottestad** [comment]
I would like to merge this, since we at least have a fix for one of the edge cases. The others I have `@ignored`.

**Comment by jeenbroekstra** [comment]
@hmottestad I may spend another hour or so today on a fix. I'll merge when I'm done either way.

File: core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java (hunk 2 above)
Comment (jeenbroekstra) [reply core/rio/turtle/src/test/java/org/eclipse/rdf4j/rio/turtle/TurtleWriterTest.java:370]: > I think the output is incorrect. The pretty printer and bnode inliner should return the same data if you first output n-triples, then pretty print and inline it, then convert it to n-triples again

My point is that effectively, it does.

**Comment by jeenbroekstra** [comment]
I figured out that the original big failing test is also caused by what is effectively a malformed list structure: there are places where list nodes have _multiple_ rdf:rest properties, which is not valid. 

I've pushed a fix that does an extensive check before attempting to use list shorthand syntax, to make sure the list is not malformed. If it is malformed, it falls back to "normal" bnode inlining, not using list shorthand. This seems to address all corner cases I could come up with in additional tests, as well as the original large test case.

**Comment by hmottestad** [comment]
That's awesome Jeen. I'm going to try to figure out where I was producing the malformed lists from.
